### PR TITLE
Add API Rate limit + security improvements

### LINKED ISSUE #46: Limit API access rate + Security Fixes
According to LGTM analysis, we are performing database accesses without limiting the API calls. After some research, I found that the most used (and it seems to be an alive project, so it is updated) is express-rate-limit (https://www.npmjs.com/package/express-rate-limit).

The alternatives to store the access cache are the node process internal memory, redis, memcached and mongo.

As we might deploy the server in a cluster in the future, across multiple node processes in parallel, it's important to store the accesses in an external cache.  redis and memcached are optimized for reading, and it seems the implementation is quite easy, so I recommend using the redis approach, by adding this package as well: https://www.npmjs.com/package/rate-limit-redis.

Additionally, after looking at the express recommendations for security (https://expressjs.com/en/advanced/best-practice-security.html), we should adopt some of them (some are already in place, yay) such as the helmet plugin (https://expressjs.com/en/advanced/best-practice-security.html#use-helmet)

### PR DESCRIPTION
**Pull request by imnotteixeira**
Closes #46 

Choose to use the mongo storage approach, for the reasons described in the issue.

Also added the helmet dependency, as recommended in the express docs, linked in the issue.

### CODE REVIEW

File: src/loaders/express.js [hunk 1]
```diff
@@ -34,6 +41,28 @@ module.exports = (app) => {
     app.use(passport.initialize());
     app.use(passport.session());
 
+    const mongo_uri = config.db_uri || `mongodb://${config.db_host}:${config.db_port}/${config.db_name}`;
```
Comment (GuilhermeJSilva) [inline src/loaders/express.js]: We have 2 separate definitions of this, even tough they are equal it would be best to have only one so if it ever needs be changed it would always stay consistent

File: src/loaders/express.js [hunk 2]
```diff
@@ -34,6 +41,28 @@ module.exports = (app) => {
...
+            message: {
+                error: "Too many requests issued from this IP, please try again after a while",
+            },
+        });
+
+        // Adds rate limit to all routes
+        app.use(api_rate_limiter);
+    }
+
+    // Adds protection to common attacks
```
Comment (GuilhermeJSilva) [inline src/loaders/express.js]: It would be better to include a link to the list of protections it provides

File: src/loaders/express.js (hunk 1 above)
Comment (imnotteixeira) [reply src/loaders/express.js]: Made it so that the uri is always available if it is provided, or if host port and db name are provided, that way, one can always access db_uri directly. There were already checks to assert that there exists either db_uri or db host/port and name. Ran the tests and it successfully connected so it should be fine.

**Comment by imnotteixeira** [comment]
Fixed conflicts with `develop`, should be ready now.

File: src/loaders/express.js [hunk 3]
```diff
@@ -35,6 +42,29 @@ module.exports = (app) => {
     app.use(passport.initialize());
     app.use(passport.session());
 
+    const mongo_uri = config.db_uri;
+
+    // Set the API call rate limit only on production
+    if (process.env.NODE_ENV === "production") {
```
Comment (miguelpduarte) [inline src/loaders/express.js]: We need to make sure that the servers in production are actually launched with `NODE_ENV=production` because of this. Commenting here so that I don't forget to quickly check this.

File: src/config/env.js [hunk 4]
```diff
@@ -20,3 +20,12 @@ module.exports = Object.freeze({
...
     admin_password: process.env.ADMIN_PASSWORD,
 });
+
+module.exports = Object.freeze({
+    ...config,
+    db_uri: config.db_uri ||
+        (config.db_host
+            && config.db_port
+            && config.db_name
+            && `mongodb://${config.db_host}:${config.db_port}/${config.db_name}`),
```
Comment (miguelpduarte) [inline src/config/env.js]: The uri that was being used in the mongoose loader did not use the db_name directly, but instead used it in the options for the connection. Please change this so that the db_name is not used here.

Uses that desire to know the db_name shall fetch it as they need to, but db_uri should just be the uri to the mongodb database.

File: src/loaders/express.js (hunk 1 above)
Comment (miguelpduarte) [reply src/loaders/express.js]: The problem is not in connecting to the tests, as they provide db_host and db_port. The problem is that the new changes would probably not work in production, in which the mongodb cluster uri is provided :sweat_smile:

I think the suggested changes solve this though.

File: src/loaders/express.js (hunk 3 above)
Comment (miguelpduarte) [reply src/loaders/express.js]: `niployments` runs `Dockerfile-prod`. `nijobs-be`'s `Dockerfile-prod` runs `npm run prod`. However, `npm run prod` does not set `NODE_ENV` to `production`.

So, if you don't mind, this needs to be set in package.json for this PR to be effective :slightly_smiling_face:

Comment (miguelpduarte) [reply src/loaders/express.js]: (Furthermore, we should really update the node versions in the dockerfiles, as they are being used for production, CI and etc. Creating an issue to track that now)

File: src/loaders/express.js [hunk 5]
```diff
@@ -8,6 +8,13 @@ const apiRoutes = require("../api");
 const config = require("../config/env");
 const { dbHandler } = require("../api/middleware/errorHandler");
 
+const helmet = require("helmet");
+
+const RateLimit = require("express-rate-limit");
+const MongoStore = require("rate-limit-mongo");
+
+const API_REQUEST_TIME_WINDOW_MS = 15 * 60 * 1000; // 15 minutes
+const API_MAX_REQUESTS_PER_WINDOW = 100;
```
[the remark this answers is not in the document]
Comment (imnotteixeira) [reply src/loaders/express.js]: Yeah, I don't know if those values are well thought, they are per IP, not in total, I think I got those from some default I found, but they can be changed

File: src/config/env.js (hunk 4 above)
Comment (imnotteixeira) [reply src/config/env.js]: There is no case where you need the URI without the db_name, I updated the mongo connection call to pass the complete URI instead of a db_name param in the options which has the same effect. The other cases always use the complete URI. Which makes sense, in the way they are connecting to a specific db instance, not to the server where the db is located

Comment (miguelpduarte) [reply src/config/env.js]: I believe that we found out that this is not the case for the test environment :sweat_smile:

File: src/loaders/express.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (miguelpduarte) [reply src/loaders/express.js]: Okay, I guess we can roll with this for now and later on see if changes are necessary :slightly_smiling_face:

**Comment by imnotteixeira** [comment]
Hey @miguelpduarte, just tested this and pushed the solution for the db_uri problem:
Basically, db_uri will be set (even if not provided), however, for example on the test scenario, if db_name is overriden, it will use that (i.e. if uri is mongodb:/<something>/db_name_1, and the .env.test sets db name to db_name_2, the connection will use db_name_2 even though the uri already has another db name in it)

**Review by GuilhermeJSilva — DISMISSED** [review]

File: src/loaders/express.js (hunk 3 above)
Comment (miguelpduarte) [reply src/loaders/express.js:46]: Resolving as this was tackled

**Comment by miguelpduarte** [comment]
I need to give this a new good look, as it has been a while so this should either just be closed or merged instead of being eternally zombie. @imnotteixeira can you please test with a URI like the ones used in staging? (the cluster ones) I think that there may still be some issues with that format when running.

File: src/config/env.js [hunk 6]
```diff
@@ -1,22 +1,27 @@
...
     db_port: process.env.DB_PORT,
-    db_uri: process.env.DB_URI,
+    db_name: process.env.DB_NAME,
+    db_uri: process.env.DB_URI ||
+    (
+        process.env.DB_HOST
+            && process.env.DB_PORT
+            && process.env.DB_NAME
+            && `mongodb://${process.env.DB_HOST}:${process.env.DB_PORT}/${process.env.DB_NAME}`
+    ),
```
Comment (miguelpduarte) [inline src/config/env.js:14]: If we are to go with this solution we should just make a method for this, I think it would be simpler or at least abstract from the rest of the configuration.

File: src/config/env.js [hunk 7]
```diff
@@ -1,22 +1,27 @@
...
-
     god_token: process.env.GOD_TOKEN,
-
     test_log_requests: JSON.parse(process.env.TEST_LOG_REQUESTS),
-
     admin_email: process.env.ADMIN_EMAIL,
     admin_password: process.env.ADMIN_PASSWORD,
 });
+
+module.exports = config;
```
Comment (miguelpduarte) [inline src/config/env.js:27]: Rationale for this change? I liked the `module.exports =` approach and don't really see reasons to change

File: src/loaders/express.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (miguelpduarte) [reply src/loaders/express.js:17]: Going back to this, I think we should increase this a bit. Especially given the idea to have infinite scrolling on the search page. Scrolling quickly with a lot of entries might easily go over 100 requests in 15 minutes, given that the person might refresh the page and then go back to browsing for a bit.
However, like you said at the time - these values are heavily based on speculation, since we don't have real user data for now. As such we probably can just go with any value and adapt later. I just think it might be better to overshoot than undershoot 😛

**Review by miguelpduarte — CHANGES_REQUESTED** [review]
I think we might be able to quickly move this forward.
By the way, I'm concerned that the rate-limiting might impact the free tier of storage we have going for the current MongoDB hosting solution. Can you provide any insight relative to this? We should probably be careful to not go over the quota.

File: src/config/env.js (hunk 7 above)
Comment (imnotteixeira) [reply src/config/env.js:27]: I'm actually not sure why it changed, maybe I was exporting additional things and changed the structure, but it can be reverted.

File: src/loaders/express.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (imnotteixeira) [reply src/loaders/express.js:17]: I honestly think this is a good starting point since for the search it will probably be only 4-5 reloads (We won't have 1000 active offers in the near future - I hope we reach it someday, though :P)

I don't oppose changing the limit, it's just important to have one, really. But I don't think this will cause us problems for now.

**Comment by imnotteixeira** [comment]
> I need to give this a new good look, as it has been a while so this should either just be closed or merged instead of being eternally zombie. @imnotteixeira can you please test with a URI like the ones used in staging? (the cluster ones) I think that there may still be some issues with that format when running.

I don't have access to the staging .env, but if it is a complete URI, it should work as per 

> Hey @miguelpduarte, just tested this and pushed the solution for the db_uri problem:
Basically, db_uri will be set (even if not provided), however, for example on the test scenario, if db_name is overriden, it will use that (i.e. if uri is mongodb://db_name_1, and the .env.test sets db name to db_name_2, the connection will use db_name_2 even though the uri already has another db name in it)

**Comment by miguelpduarte** [comment]
The thing is that the URI has the specific requirement of being a cluster URI and having to specify the dbname separately, as mentioned elsewhere in this PR (can't really search it atm). We cannot use the URI to specify the dbname AFAIK, as when we tried that with the MongoDB Atlas deployment it no longer worked.

**Comment by imnotteixeira** [comment]
> The thing is that the URI has the specific requirement of being a cluster URI and having to specify the dbname separately, as mentioned elsewhere in this PR (can't really search it atm). We cannot use the URI to specify the dbname AFAIK, as when we tried that with the MongoDB Atlas deployment it no longer worked.

Just tested this and basically, it first looks at the URI, and if present it uses it. Additionally, if there is a `DB_NAME` present it will override the one in the URI.

Finally, if no URI is present, DB_HOST and DB_PASS are used, together with the DB_NAME, for the tests use case, for example, where localhost is used.

Tl;dr, it works.

File: src/config/env.js [hunk 8]
```diff
@@ -1,22 +1,26 @@
 require("dotenv-flow").config();
 
+const generateDBUriFromProps = () => {
```
Comment (miguelpduarte) [inline src/config/env.js:3]:
```suggestion
const generateDBUriFromEnv = () => {
```
Too much React? 😛

File: src/loaders/express.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (miguelpduarte) [reply src/loaders/express.js:17]: Resolving this for now though. Maybe we should create an issue to further analyze and update this later, though.

File: src/config/env.js (hunk 8 above)
Comment (imnotteixeira) [reply src/config/env.js:3]: No React is too much React.

**Review by miguelpduarte — APPROVED** [review]
👍
